keps/sig-node/1948-deallocate-dp-api/README.md

```diff
@@ -0,0 +1,846 @@
+
+- `Deallocate`:
+
+  - For use with complex device plugins that require tracking the state of their
+  devices and learning when they are no longer in use. Multi modal devices can
+  operate in more than one mode of operation, and thus have to be advertised by
+  the device plugin as two separate devices, and the device plugin has to take
+  care to stop advertising a device when its being used in the other mode. An
+  example of multi modal devices is also using FPGAs in the following 2 modes:
+    - Use the entire FPGA as a device
+    - Split the FPGA between multiple users, essentially advertising one FPGA as
+    multiple smaller FPGAs.
+    - A device plugin needs to know when a full FPGA will stop being used so it
+    can go back to advertise the FPGA partitions, and vice versa.
+  - To maintain the same logical splitting of `Allocate` and `PreStartContainer`
+
+### Goals
+
+- Add the `PostStopContainer` and `Deallocate` API calls to the device plugin
+API.
+- Make the new added API calls optional, as they are not needed for all devices.
+- Maintain compatibility with existing device plugins.
+
+### Non-Goals
+
+- Make any modifications to the main API calls of the device plugin API.
+- Make changes specific to one type of devices.
+
+## Use Cases
+
+- Power management: in case of accelerators that are not always depending on
+controlling process from userland side (e.g. FPGA which has external
+network/GPIO interfaces) it is possible that after stopping of Pod where device
+was initially allocated, accelerator part of the code will be still processing
+data on external interfaces and consume significant amount of power. As we don't
+know when it will be next time allocated, it will be better to reset accelerator
+device and put it into power saving mode immediately after workload that
+requested it is terminated.
+
+- Security: HW-accelerators often have internal memory and processing data not
+always cleared after processing. Doing reset/memory clean-up on next Allocate
+call means that between two points of time (termination of workload and next
+Allocate) some confidential information can be still present on the device (e.g.
+IPsec encryption keys, for crypto accelerators). It is better to have ability to
+cleanup device memory as soon as workload that requested it finishes.
+
+- Virtualization: HW-accelerators can be used and resued in several ways in a
+VM.
+  - A device bound to VFIO and passthrough'ed into the VM
+  - A device can be configured as an mediated device and used in a VM
+  - A device can be configured with SR-IOV and specific VFs or a PF can be used in
+a VM
+A device may support all of those modi or a subset. Depending on what is needed
+for the workload a device may be bound to VFIO for performance reasons e.g. and
+released/deallocated for usage in a different workload as an mediated device.
+
+- Bind/Unbind: Instead of pre-configuring the devices on the node with the
```

Comment:
Additional examples for minimizing pre-configuration : Creating mediated devices (like vgpu) at the time of `Allocate` and deleting them at the time of `Deallocate`, thus allowing for a single physical gpu to be used either as both passthrough GPU or vGPU by configuring it accordingly at the time of allocation and resetting the device at the time of deallocation.